src/main/java/com/google/firebase/auth/internal/CryptoSigners.java

```diff
@@ -0,0 +1,166 @@
+    }
+
+    @Override
+    public byte[] sign(byte[] payload) throws IOException {
+      String encodedUrl = String.format(IAM_SIGN_BLOB_URL, serviceAccount);
+      HttpResponse response = null;
+      String encodedPayload = BaseEncoding.base64().encode(payload);
+      Map<String, String> content = ImmutableMap.of("bytesToSign", encodedPayload);
+      try {
+        HttpRequest request = requestFactory.buildPostRequest(new GenericUrl(encodedUrl),
+            new JsonHttpContent(jsonFactory, content));
+        request.setParser(new JsonObjectParser(jsonFactory));
+        request.setResponseInterceptor(interceptor);
+        response = request.execute();
+        SignBlobResponse parsed = response.parseAs(SignBlobResponse.class);
+        return BaseEncoding.base64().decode(parsed.signature);
+      } finally {
+        if (response != null) {
+          try {
+            response.disconnect();
+          } catch (IOException ignored) {
+            // Ignored
+          }
+        }
+      }
+    }
+
+    @Override
+    public String getAccount() {
+      return serviceAccount;
+    }
+  }
+
+  public static class SignBlobResponse {
+    @Key("signature")
+    private String signature;
+  }
+
+  /**
+   * Initializes a {@link CryptoSigner} instance for the given Firebase app. Follows the protocol
+   * documented at go/firebase-admin-sign.
+   */
+  static CryptoSigner getCryptoSigner(FirebaseApp firebaseApp) throws IOException {
+    GoogleCredentials credentials = ImplFirebaseTrampolines.getCredentials(firebaseApp);
+
+    // If the SDK was initialized with a service account, use it to sign bytes.
+    if (credentials instanceof ServiceAccountCredentials) {
+      return new ServiceAccountCryptoSigner((ServiceAccountCredentials) credentials);
+    }
+
+    FirebaseOptions options = firebaseApp.getOptions();
+    HttpRequestFactory requestFactory = options.getHttpTransport().createRequestFactory(
+        new FirebaseRequestInitializer(firebaseApp));
+    JsonFactory jsonFactory = options.getJsonFactory();
+
+    // If the SDK was initialized with a service account email, use it with the IAM service
+    // to sign bytes.
+    String serviceAccount = options.getServiceAccount();
+    if (!Strings.isNullOrEmpty(serviceAccount)) {
+      return new IAMCryptoSigner(requestFactory, jsonFactory, serviceAccount);
+    }
+
+    // If the SDK was initialized with some other credential type that supports signing
+    // (e.g. GAE credentials), use it to sign bytes.
+    if (credentials instanceof ServiceAccountSigner) {
+      return new ServiceAccountCryptoSigner((ServiceAccountSigner) credentials);
+    }
+
+    // Attempt to discover a service account email from the local Metadata service. Use it
+    // with the IAM service to sign bytes.
+    HttpRequest request = requestFactory.buildGetRequest(new GenericUrl(METADATA_SERVICE_URL));
+    request.getHeaders().set("Metadata-Flavor", "Google");
+    HttpResponse response = request.execute();
```

Comment:
Do we have to to call `response.disconnect()` here as well?

Reply:
Done